### ISSUE TITLE
Unrecognized JSON field when working with New-HPOVServerProfileTemplate with -SANStorage set

### ISSUE THREAD

**Issue by tknightblc** [HewlettPackard/POSH-HPOneView#99]
Chris,

All my server profile templates are working with latest build except two where I create SAN volumes I am getting an unrecognized JSON field error when using New-HPOVServerProfileTemplate with -SANStorage set - specifically something to do with the --StorageVolume $attachVolumes3 below. I think I built these correctly, can you please take a look when you get a chance?

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> Get-HPOVVersion

PowerShell Library Version : 2.0.312.0
Library Path               : C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $PSVersionTable

Name  : PSVersion
Value : 4.0

Name  : WSManStackVersion
Value : 3.0

Name  : SerializationVersion
Value : 1.1.0.1

Name  : CLRVersion
Value : 4.0.30319.42000

Name  : BuildVersion
Value : 6.3.9600.16406

Name  : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0}

Name  : PSRemotingProtocolVersion
Value : 2.2

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> # Get networks ready
$MA = Get-HPOVNetwork -name "ESX-Mgmt-A" -ApplianceConnection $Connection1
$MB = Get-HPOVNetwork -name "ESX-Mgmt-B" -ApplianceConnection $Connection1
$vMA = Get-HPOVNetwork -name "ESX-Vmotion-A" -ApplianceConnection $Connection1
$vMB = Get-HPOVNetwork -name "ESX-Vmotion-B" -ApplianceConnection $Connection1
$OBNetSetA = Get-HPOVNetworkSet -name "ESX-Outbound-A" -ApplianceConnection $Connection1
$OBNetSetB = Get-HPOVNetworkSet -name "ESX-Outbound-B" -ApplianceConnection $Connection1
$SANA = Get-HPOVNetwork -name "FC-SAN-A" -ApplianceConnection $Connection1
$SANB = Get-HPOVNetwork -name "FC-SAN-B" -ApplianceConnection $Connection1

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $FlexNic1 = New-HPOVProfileConnection -id 1 -connectionType Ethernet -requestedBW 1000 -network $MA -ApplianceConnection $Connection1
$FlexNic2 = New-HPOVProfileConnection -id 2 -connectionType Ethernet -requestedBW 1000 -network $MB -ApplianceConnection $Connection1
$FlexNic3 = New-HPOVProfileConnection -id 3 -connectionType FibreChannel -requestedBW 4000 -network $SANA -ApplianceConnection $Connection1
$FlexNic4 = New-HPOVProfileConnection -id 4 -connectionType FibreChannel -requestedBW 4000 -network $SANB -ApplianceConnection $Connection1
$FlexNic5 = New-HPOVProfileConnection -id 5 -connectionType Ethernet -requestedBW 3000 -network $OBNetSetA -ApplianceConnection $Connection1
$FlexNic6 = New-HPOVProfileConnection -id 6 -connectionType Ethernet -requestedBW 3000 -network $OBNetSetB -ApplianceConnection $Connection1
$FlexNic7 = New-HPOVProfileConnection -id 7 -connectionType Ethernet -requestedBW 2000 -network $vMA -ApplianceConnection $Connection1
$FlexNic8 = New-HPOVProfileConnection -id 8 -connectionType Ethernet -requestedBW 2000 -network $vMB -ApplianceConnection $Connection1
$conListA = @($FlexNic1, $FlexNic2, $FlexNic3, $FlexNic4, $FlexNic5, $FlexNic6, $FlexNic7, $FlexNic8)

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> # Create storage attachements for templates
$BootVolume1 = Get-HPOVStorageVolume Boot-BL660-1 | New-HPOVProfileAttachVolume -volumeid 1
$SharedVolume1 = Get-HPOVStorageVolume VMFS-1 | New-HPOVProfileAttachVolume -volumeid 2
$SharedVolume2 = Get-HPOVStorageVolume VMFS-2 | New-HPOVProfileAttachVolume -volumeid 3
$attachVolumes1 = @($BootVolume1,$SharedVolume1,$SharedVolume2)
$BootVolume2 = Get-HPOVStorageVolume Boot-BL660-2 | New-HPOVProfileAttachVolume -volumeid 1
$attachVolumes2 = @($BootVolume2,$SharedVolume1,$SharedVolume2)
$attachVolumes3= @($SharedVolume1,$SharedVolume2)

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> # Set BIOS settings for max performance and Administrator contact
$Gen8BiosSettings = @(@{id=210;value=3},@{id=140;value=3},@{id=108;value="Thomas Knight"},@{id=110;value="585-451-6254"})
$Gen9BiosSettings = @(@{id="PowerProfile";value="MaxPerf"},@{id="AdminName";value="Thomas Knight"},@{id="AdminPhone";value="585-451-6254"})

#get enclosure group

$eg = Get-HPOVEnclosureGroup "ESX-Enclosure"

# get server hardware types

$bl460Gen9SHT = Get-HPOVServerHardwareTypes -name "BL460c Gen9 1"
$bl460Gen8SHT = Get-HPOVServerHardwareTypes -name "BL460c Gen8 1"
$bl660Gen9SHT = Get-HPOVServerHardwareTypes -name "BL660c Gen9 1"

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> # Do settings for local disk 
$logicaldisk = New-HPOVServerProfileLogicalDisk -Name "My local disk" -RAID RAID1 -NumberofDrives 2 -DriveType SAS

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $attachVolumes3

id                     : 2
volumeUri              : /rest/storage-volumes/B562B972-3D64-49C9-99B8-7E454531B1A7
volumeStoragePoolUri   : /rest/storage-pools/C2D85619-40E0-4221-8E3C-DB1EBABC7EB4
volumeStorageSystemUri : /rest/storage-systems/TXQ1000307
lunType                : Auto
storagePaths           : {}
ApplianceConnection    : @{Name=tk1-ov200-dcsdemo.advizex.net; ConnectionId=1}

id                     : 3
volumeUri              : /rest/storage-volumes/B728B5C0-CFDA-4161-90E0-5F040F9306C0
volumeStoragePoolUri   : /rest/storage-pools/C2D85619-40E0-4221-8E3C-DB1EBABC7EB4
volumeStorageSystemUri : /rest/storage-systems/TXQ1000307
lunType                : Auto
storagePaths           : {}
ApplianceConnection    : @{Name=tk1-ov200-dcsdemo.advizex.net; ConnectionId=1}

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> New-HPOVServerProfileTemplate -name "A BL460 Gen9 ESX Template with SAN" -description "Template for ESX host-BL460 Gen9 with shared SAN storage definitions" -firmware -baseline "HP Service Pack For ProLiant" -serverHardwareType $bl460Gen9SHT -enclosureGroup $eg -connections $conListA –manageBoot -pxeBootPolicy Auto -bios -biossettings $Gen9BiosSettings -localstorage -Initialize -ControllerMode RAID -LogicalDisk $logicaldisk -SANStorage -HostOStype VMware -StorageVolume $attachVolumes3 -ApplianceConnection $Connection1 | Wait-HPOVTaskComplete

New-HPOVServerProfileTemplate : Unrecognized JSON field. The field named ApplianceConnection near line 1, column 3,282 is not a valid field in this JSON object.
At line:1 char:1
+ New-HPOVServerProfileTemplate -name "A BL460 Gen9 ESX Template with SAN" -descri ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Send-HPOVRequest:String) [New-HPOVServerProfileTemplate], InvalidOperationException
    + FullyQualifiedErrorId : InvalidAuthOperation,New-HPOVServerProfileTemplate
 

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $error | % { Format-List -InputObject $_ * } 

PSMessageDetails      : 
Exception             : System.InvalidOperationException: Unrecognized JSON field. The field named ApplianceConnection near line 1, column 3,282 is not a valid field in 
                        this JSON object.
                           at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)
TargetObject          : Send-HPOVRequest
CategoryInfo          : InvalidOperation: (Send-HPOVRequest:String) [Send-HPOVRequest], InvalidOperationException
FullyQualifiedErrorId : InvalidAuthOperation,Send-HPOVRequest
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Send-HPOVRequest<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 5486
                        at New-HPOVServerProfileTemplate<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 41477
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

PSMessageDetails      : 
Exception             : System.Management.Automation.MethodInvocationException: Exception calling "GetResponse" with "0" argument(s): "The remote server returned an 
                        error: (400) Bad Request." ---> System.Net.WebException: The remote server returned an error: (400) Bad Request.
                           at System.Net.HttpWebRequest.GetResponse()
                           at CallSite.Target(Closure , CallSite , HttpWebRequest )
                           --- End of inner exception stack trace ---
                           at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
                           at lambda_method(Closure , Object[] , StrongBox`1[] , InterpretedFrame )
TargetObject          : 
CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
FullyQualifiedErrorId : WebException
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Send-HPOVRequest<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 5085
                        at New-HPOVServerProfileTemplate<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 41477
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $error | % { $_.ScriptStackTrace } 
at Send-HPOVRequest<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 5486
at New-HPOVServerProfileTemplate<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 41477
at <ScriptBlock>, <No file>: line 1
at Send-HPOVRequest<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 5085
at New-HPOVServerProfileTemplate<Process>, C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1: line 41477
at <ScriptBlock>, <No file>: line 1

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy>

**Comment by ChrisLynchHPE** [comment]
Looks like the `ApplianceConnection` property we use for object tracking isn't being removed from the Volume object.  As a workaround until a fix is released, you can use the following:

```PowerShell
$BootVolume1 = Get-HPOVStorageVolume Boot-BL660-1 | New-HPOVProfileAttachVolume -volumeid 1 | Select * -ExcludeProperty ApplianceConnection
$SharedVolume1 = Get-HPOVStorageVolume VMFS-1 | New-HPOVProfileAttachVolume -volumeid 2 | Select * -ExcludeProperty ApplianceConnection
$SharedVolume2 = Get-HPOVStorageVolume VMFS-2 | New-HPOVProfileAttachVolume -volumeid 3 | Select * -ExcludeProperty ApplianceConnection
$BootVolume2 = Get-HPOVStorageVolume Boot-BL660-2 | New-HPOVProfileAttachVolume -volumeid 1 | Select * -ExcludeProperty ApplianceConnection
```
That will remove the `ApplianceConnection` property from the object, and allow you to create the SPT.

**Comment by tknightblc** [comment]
Ok thanks much Chris!

**Comment by tknightblc** [comment]
Chris,

Ok that worked but new issue – evidently you cannot have –LunIDType Auto with a storage template, says I must use manual. So I tried to set -LunIDTpe Manual with –LunID 2 and get this error:

PS C:\Users\tknight\Desktop\OneView\HP OV 2.0 DCS\Toms PS scripts - 2.0 working copy> $SharedVolume1 = Get-HPOVStorageVolume VMFS-1 | New-HPOVProfileAttachVolume -volumeid 2 -LunIdType Manual -LunID 2 | Select * -ExcludeProperty ApplianceConnection
Exception setting "lun": "The property 'lun' cannot be found on this object. Verify that the property exists and can be set."
At C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1:44066 char:21
+                     $volumeAttachment.lun = $LunID
+                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], SetValueInvocationException
    + FullyQualifiedErrorId : ExceptionWhenSetting
Exception setting "lun": "The property 'lun' cannot be found on this object. Verify that the property exists and can be set."
At C:\Windows\system32\WindowsPowerShell\v1.0\Modules\HPOneView.200\HPOneView.200.psm1:44253 char:5
+                 $volumeAttachment.lun     = $LunID
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], SetValueInvocationException
    + FullyQualifiedErrorId : ExceptionWhenSetting

I can enter this as a new issue with the debug info if you want me to…




From: Thomas Knight
Sent: Tuesday, April 19, 2016 2:44 PM
To: 'HewlettPackard/POSH-HPOneView' <reply@reply.github.com>; HewlettPackard/POSH-HPOneView <POSH-HPOneView@noreply.github.com>
Subject: RE: [HewlettPackard/POSH-HPOneView] Unrecognized JSON field when working with New-HPOVServerProfileTemplate with -SANStorage set (#99)

Ok thanks much Chris!

**Comment by ChrisLynchHPE** [comment]
Go ahead and open another issue.  The `lun` property is missing from the object, and looks to be a regression issue (although I don’t know why ATM).

**Comment by tknightblc** [comment]
Done

**Comment by ChrisLynchHPE** [comment]
This is fixed in [Release 2.0.368.0](https://github.com/HewlettPackard/POSH-HPOneView/releases/tag/v2.0.368.0).

Closing.
